Feature/auth service/creating dto and controller

Goodwin999 opened the pull request:
created a controller and DTO, designated annotations for generating swagger documentation;
User entity: In the DB we store only id, username, password, role (as in your code).
DTO: Contains all fields for the API, including firstName, lastName, phone (but they are not mapped to the entity).

Goodwin999 committed:
feat(auth): created user dto

Goodwin999 committed:
feat(auth): created a controller

Goodwin999 committed:
feat(auth): Merge dev to get User entity

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,30 @@
+    @Operation(summary = "Регистрация", description = "Создает нового пользователя")
+    @PostMapping("/register")
+    public UserResponse register(@RequestBody RegisterRequest request) {
+        return new UserResponse();
```
Это все что должен делать метод контроллера? вернуть пустое dto?
Мне кажется, что этого не полный функционал )
Более того по документации этот метод вообще ничего не должен возвращать

Goodwin999 replied:
![image](https://github.com/user-attachments/assets/8afee146-05e8-41ef-a745-6afd07f531a6)

belousoveu replied:
Посмотри документацию именно по своим эндпойнтам.
твои эндпойнты ничего возвращать не должны.
И, если ты еще не заметил - мы не делаем наш проект по шаблону СкайПро. Не надо делать бестолковых вещей, типа пустых контроллеров, которые потом нужно переделывать.

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,30 @@
+    @Operation(summary = "Логин", description = "Аутентификация и получение JWT-токена")
+    @PostMapping("/login")
+    public JwtResponse login(@RequestBody LoginRequest request) {
+        return new JwtResponse();
```
то же самое что и выше.

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/dto/request/LoginRequest.java:
```diff
@@ -0,0 +1,15 @@
+@Data
+public class LoginRequest {
+    @NotBlank
+    @Size(min = 4, max = 32)
```
Объясни параметры валидации.
У меня ощущение, что тут должны быть другие критерии валидации

Goodwin999 replied:
![image](https://github.com/user-attachments/assets/26be0748-5c0f-41fc-808e-c36dd9aa5d59)

belousoveu replied:
Слава, у нас логином является электронная почта. сделай проверку, чтобы логин соответствовал формату электронной почты.
И кстати, сможешь написать адрес электронной почты состоящий из 4 символов? )

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/dto/request/RegisterRequest.java:
```diff
@@ -0,0 +1,30 @@
+public class RegisterRequest {
+    @NotBlank
+    @Size(min = 4, max = 32)
+    private String username;
```
Неверные критерии валидации

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/dto/request/RegisterRequest.java:
```diff
@@ -0,0 +1,30 @@
+
+    @NotBlank
+    @Size(min = 8, max = 16)
+    private String password;
```
Давай введем требование к паролю
мин - 8 символов согласен
допускаются только буквы латинского алфавита и цифры
Пароль должен содержать как минимум одну прописную букву, минимум одну заглавную букву, минимум одну цифру

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/dto/request/RegisterRequest.java:
```diff
@@ -0,0 +1,30 @@
+    @Size(min = 8, max = 16)
+    private String password;
+
+    @NotBlank
```
Вместо `@NotBlank` - `@NotNull`
NotBlank тут бесполезна, поскольку дублируется следующей аннотацией

Goodwin999 replied:
Не совсем понял почему, и на какие поля конкретно ты говоришь? Емейл?
Я почитал разницу между ними, NotBlank же как раз для строк, а у нас все в строках находится, но я поменял как ты сказал

belousoveu replied:
`@NotBlank` и `@NotNull` работают по разному
если стоит аннотация NotBlank  - а на проверку отправят null, то валидатор не выдаст ошибки. 
Ты уверен, что именно такой логики ты ждешь от своего кода?

Зато потом возникнет ошибка на этапе записи в БД, потому что в таблице стоит условие not null у этого поля.
Вообще все условия, которые прописаны при создании таблиц в файле миграции должны точно соответствовать аннотациям hibernate, чтобы тот эффективно работал

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/dto/request/RegisterRequest.java:
```diff
@@ -0,0 +1,30 @@
+    @Size(min = 2, max = 16)
+    private String firstName;
+
+    @NotBlank
```
аналогично

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/dto/request/RegisterRequest.java:
```diff
@@ -0,0 +1,30 @@
+    private String lastName;
+
+    @NotBlank
+    @Pattern(regexp = "\\+7\\s?\\(?\\d{3}\\)?\\s?\\d{3}-?\\d{2}-?\\d{2}")
```
+

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/dto/request/RegisterRequest.java:
```diff
@@ -0,0 +1,30 @@
+    @Size(min = 2, max = 16)
+    private String lastName;
+
+    @NotBlank
```
Может быть пустым. Почему не может?

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/dto/request/response/JwtResponse.java:
```diff
@@ -0,0 +1,8 @@
+import lombok.Data;
+
+@Data
+public class JwtResponse {
```
Зачем нужен класс  с одним текстовым полем?
Почему просто не использовать строку?
обоснуй применение отдельного класса dto для JWT

Goodwin999 replied:
Требуется уточнение, не могу сам разобраться: 
пришел к тому, что надо удалить текущее поле, вместо него добавить 
private String accessToken;
private String refreshToken;
+ возможно expiresIn, tokenType;

belousoveu replied:
Да, ты все правильно написал.
Тебе в ответ на запрос авторизации клиента нужно вернуть ему два токена в заголовке ответа
access и refresh токены.
Соответственно тебе нужно их передать в контроллер, потому что он формирует ответ поэтому в этом классе должно быть 2 поля, оба типа String.
и поскольку оба этих поля финальные, то вообще я бы тут использовал класс record для этого (но это не является критичным)
Почему, на мой взгляд лучше record - проще читать код. Сразу понимаешь, поведение этого объекта и как он используется дальше в коде.

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/dto/request/response/UserResponse.java:
```diff
@@ -0,0 +1,13 @@
+import lombok.Data;
+
+@Data
+public class UserResponse {
```
Для чего нужен этот класс? где он будет использоваться?
Я не вижу применения для этого класса в таком виде

Goodwin999 replied:
Требуется уточнение, не могу сам разобраться:
Вроде мы обсуждали тогда на созвоне, что в сущности у меня только данные, которые я храню в БД, а остальное будет в дто для безопасной передачи в другие сервисы(без пароля)
Пока не могу понять, оставлю без изменений эти вещи до твоего ответа.

belousoveu replied:
Данные в при регистрации пользователя, которые не хранятся в твоей БД ты дальше передаешь через Kafak. Соответственно тебе нужно dto для формирования сообщения в Kafka.
Посмотри какие поля предусмотрены для этого.
На мой взгляд поле role тут лишнее. Роль не должна передаваться другим сервисам, поскольку она им в данном контексте не нужна.
Роль пользователя другим сервисам нужна для авторизации запроса и они соответственно должны получать ее вместе с http-запросом в JWT (вот в payload JWT его надо включить обязательно)

belousoveu reviewed (changes requested):
Слава, я тебе как то уже писал, повторю.
Это здорово, что ты умеешь пользоваться нейросетью (абсолютно без иронии, это и правда очень полезный навык)
Но ты не должен просто копировать код нейросети. Ты должен пропускать его через себя и анализировать и понимать каждую строчку. Почему именно так, не по-другому. Как можно сделать иначе, какие варианты.
Второе замечание - задачи надо делать чтобы они зависели только от уже выполненных задач. Контроллер у тебя зависит от сервиса, по определению. Чтобы его писать - нужно сначала написать сервис значит. чтобы делать сервис - нужно сначала написать репозиторий, потому что сервис от репозитория зависит.
А вот репозиторий зависит от Entity, которая у тебя уже готова. Вот с репозитория и надо начинать. - далее сервис - далее контроллер.
А сейчас то что ты написал - контроллером назвать нельзя - потому что свои задачи он не выполняет

Goodwin999 committed:
feat(auth):
-added email validation;
-added password validation;

Goodwin999 committed:
feat(auth):
updated the controller and added an endpoint for refreshing the token, I implement services in another branch;

Goodwin999 committed:
feat(auth):
added services without logic yet, it will be written later;

Goodwin999 commented:
> Слава, я тебе как то уже писал, повторю. Это здорово, что ты умеешь пользоваться нейросетью (абсолютно без иронии, это и правда очень полезный навык) Но ты не должен просто копировать код нейросети. Ты должен пропускать его через себя и анализировать и понимать каждую строчку. Почему именно так, не по-другому. Как можно сделать иначе, какие варианты. Второе замечание - задачи надо делать чтобы они зависели только от уже выполненных задач. Контроллер у тебя зависит от сервиса, по определению. Чтобы его писать - нужно сначала написать сервис значит. чтобы делать сервис - нужно сначала написать репозиторий, потому что сервис от репозитория зависит. А вот репозиторий зависит от Entity, которая у тебя уже готова. Вот с репозитория и надо начинать. - далее сервис - далее контроллер. А сейчас то что ты написал - контроллером назвать нельзя - потому что свои задачи он не выполняет

Исправил замечания, но столкнулся со сложностями, я создал уже отдельную ветку, для сервисов и мапперов.
контроллер поправил не сильно вдаваясь пока, накиную сюда же в эту ветку сервис и реализацию(без логики) 
добавил эндпоинт /refresh
по поводу необходимости в классах на которые ты указал, я написал тебе ниже.
регулярки добавил человеческие;
в той ветке для сервисов (creating-services-and-mappers)  я создал только пакет репозитория и интерфейс UserRepository;
Мне нужна подсказка, как правильнее дальше поступить, что бы не вышло конфликта.
либо я могу сделать это все в этой ветке, остальными добавлять Jwt и остальное, хотя опять же, у Jwt должен быть такой же сервис, реализация + конфиг.
Help 


p.s.
Возвращение id в контроллере прописал потому что увидел это в критериях оценки в Тз 1 этапа, если не прав, подскажи где ты прочитал то, о чем говорил ты,(что мы ничего не возвращаем)?

Goodwin999 committed:
Update JwtResponse.java

Goodwin999 committed:
Update UserResponse.java

Goodwin999 committed:
Update AuthController.java

Goodwin999 committed:
Update AuthController.java

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,57 @@
+    @ApiResponse(responseCode = "400", description = "Некорректные входные данные")
+    @ApiResponse(responseCode = "409", description = "Пользователь с таким username уже существует")
+    @PostMapping("/register")
+    public UserResponse register(
```
Почему метод возвращает `UserResponse`?
Я же уже вроде писал об этом. по api-doc этот метод ничего е возвращает
![image](https://github.com/user-attachments/assets/5ce32c1f-edef-432d-9ae8-4f289972def4)

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,57 @@
+    @ApiResponse(responseCode = "200", description = "Успешная аутентификация")
+    @ApiResponse(responseCode = "401", description = "Неверные учетные данные")
+    @PostMapping("/login")
+    public JwtResponse login(
```
Тоже самое. JwtResponse возвращается в заголовке, а не в теле ответа

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,57 @@
+    @ApiResponse(responseCode = "200", description = "Токен обновлен")
+    @ApiResponse(responseCode = "401", description = "Невалидный refresh-токен")
+    @PostMapping("/refresh")
+    public JwtResponse refreshToken(
```
Тоже самое

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/dto/request/LoginRequest.java:
```diff
@@ -0,0 +1,15 @@
+
+@Data
+public class LoginRequest {
+    @NotBlank   // Максимальная длина email по RFC - Международный стандарт формата email
```
`@NotNull`

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/dto/request/LoginRequest.java:
```diff
@@ -0,0 +1,15 @@
+@Data
+public class LoginRequest {
+    @NotBlank   // Максимальная длина email по RFC - Международный стандарт формата email
+    @Pattern(regexp = "^(?=.{1,64}@)[\\p{L}0-9_-]+(\\.[\\p{L}0-9_-]+)*@[^-][\\p{L}0-9-]+(\\.[\\p{L}0-9-]+)*(\\.\\p{L}{2,})$")
```
Есть же аннотация валидации `@Email`. Зачем изобретать велосипед?

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/dto/request/RegisterRequest.java:
```diff
@@ -0,0 +1,29 @@
+@Data
+public class RegisterRequest {
+    @NotBlank   // Максимальная длина email по RFC - Международный стандарт формата email
+    @Pattern(regexp = "^(?=.{1,64}@)[\\p{L}0-9_-]+(\\.[\\p{L}0-9_-]+)*@[^-][\\p{L}0-9-]+(\\.[\\p{L}0-9-]+)*(\\.\\p{L}{2,})$")
```
`@Email`

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/dto/request/RegisterRequest.java:
```diff
@@ -0,0 +1,29 @@
+
+@Data
+public class RegisterRequest {
+    @NotBlank   // Максимальная длина email по RFC - Международный стандарт формата email
```
`@NotNull`

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/dto/request/LoginRequest.java:
```diff
@@ -0,0 +1,15 @@
+    @Pattern(regexp = "^(?=.{1,64}@)[\\p{L}0-9_-]+(\\.[\\p{L}0-9_-]+)*@[^-][\\p{L}0-9-]+(\\.[\\p{L}0-9-]+)*(\\.\\p{L}{2,})$")
+    private String username;
+
+    @NotBlank  //От 8 до 16 символов, одну заглавную, одну строчную буквыб одну цифру, латинские буквы
```
`@NotNull`

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/service/AuthService.java:
```diff
@@ -0,0 +1,17 @@
+import com.cherkizon.auth.dto.request.response.UserResponse;
+
+public interface AuthService {
+    UserResponse register(RegisterRequest request);
```
Зачем этому методу возвращать `UserResponse`?

Goodwin999 replied:
не понял вопроса

belousoveu replied:
Ты в сигнатуре метода пишешь, что метод должен возвращать объект `UserResponse`
Кому он его должен возвращать и зачем?
Что должен делать этот метод?

Goodwin999 replied:
проверил спецификацию, буду возвращать 201 статус

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/service/AuthService.java:
```diff
@@ -0,0 +1,17 @@
+
+    JwtResponse login(LoginRequest request);
+
+    JwtResponse refreshToken(String refreshToken);
```
Думаю, что этот метод должен быть не в этом классе

Goodwin999 replied:
создал JwtService и JwtServiceImpl перенес его туда

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/service/AuthService.java:
```diff
@@ -0,0 +1,17 @@
+    JwtResponse refreshToken(String refreshToken);
+
+
+    UserResponse getUserById(Long userId);
```
Поясни для какого сценария может быть использован этот метод?
Я такого сценария не вижу

Goodwin999 replied:
Я решил, что он должен участвовать в логике валидации, существует ли уже такой пользователь, и проверяет его существование.

belousoveu replied:
Это правильно и проверка должна быть обязательна. Но метод сервиса-то такой зачем для этого?
Проверка такая выполняется внутри сервисного метода. Снаружи этот метод ты откуда собрался вызывать?
По хорошему такая проверка делается внутри метода `register()` и даже если выносить эту проверку в отдельный метод (что наверное правильно с точки зрения SRP, хотя сам этот метод будет состоять из одной строки), то в любом случае - этот метод будет `private` и в интерфейсе ему делать нечего

Goodwin999 replied:
Принял, сейчас на свежую голову только маленько пришло понимание

belousoveu reviewed (changes requested):
Нужно до конца доработать.
И дополнительные вопросы появились к сервисным методам, которые ты добавил.
Вот поэтому я говорю - не надо в одном PR решать несколько не связанных задач

Goodwin999 committed:
Update RegisterRequest.java

Goodwin999 committed:
Update LoginRequest.java

Goodwin999 committed:
Update LoginRequest.java

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/dto/request/RegisterRequest.java:
```diff
@@ -0,0 +1,29 @@
+@Data
+public class RegisterRequest {
+    @NotNull   // Максимальная длина email по RFC - Международный стандарт формата email
+    @Email(regexp = "^(?=.{1,64}@)[\\p{L}0-9_-]+(\\.[\\p{L}0-9_-]+)*@[^-][\\p{L}0-9-]+(\\.[\\p{L}0-9-]+)*(\\.\\p{L}{2,})$")
```
Слава, c `@Email` можно не прописывать регулярку. Там по умолчанию стандартная регулярка на e-mail встроена.

Goodwin999 replied:
Она такая красивая можно оставлю☹️

belousoveu replied:
Только, если не подглядывая никуда сможешь по каждому пункту рассказать что и как проверяет это регулярное выражение )
Ну. а если серьезно - то не надо оставлять.
Почему не надо - такие вещи сильно затрудняют чтение. Когда у тебя в коде написано `@Email` - сразу понятно, что используется стандартное регулярное выражение.
Когда у тебя регулярка кастомная - ты вынужден вчитываться, чтобы понять, какую дополнительную логику (отличную от стандартной) сюда вложил разработчик. Если это действительно важно и критично влияет на функционал приложения, то да, добавление кастомной регулярки оправдано. Если - это просто, чтобы показать, что ты умеешь добавлять кастомную регулярку в аннотацию - то нет, тогда это просто "визуальный мусор".

Goodwin999 replied:
Принял, устранил

Goodwin999 committed:
feat(auth):
made edits according to comments, updated maven to 3.9.9;

Goodwin999 committed:
fix(auth):
Removed the Data annotation from the record-class

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/dto/request/RegisterRequest.java:
```diff
@@ -0,0 +1,29 @@
+    @Pattern(regexp = "\\+7\\s?\\(?\\d{3}\\)?\\s?\\d{3}-?\\d{2}-?\\d{2}")
+    private String phone;
+
+    @NotBlank
```
Так а здесь `@NotNull`?

Goodwin999 replied:
Исправил

belousoveu commented on backend/auth-service/src/main/java/com/cherkizon/auth/entity/User.java:
```diff
@@ -17,7 +18,8 @@ public class User {
 
-    @Column(unique = true, nullable = false, length = 32)
+    @Column(unique = true, nullable = false, length = 254) // Максимальная длина email по RFC - Международный стандарт формата email
+    @Pattern(regexp = "^(?=.{1,64}@)[\\p{L}0-9_-]+(\\.[\\p{L}0-9_-]+)*@[^-][\\p{L}0-9-]+(\\.[\\p{L}0-9-]+)*(\\.\\p{L}{2,})$")
```
Замени на `@Email`

Goodwin999 replied:
исправил

belousoveu reviewed (changes requested):
Слава, проверь еще раз. Ты где-то поправил, где-то пропустил.
На вопросы твои я ответил

Goodwin999 committed:
fix(auth):
editing comments

Goodwin999 committed:
fix(auth):
editing comments

Goodwin999 committed:
fix(auth):
correction of annotations

belousoveu reviewed (approved):
Теперь все ок. Можно двигаться дальше.
Советую все-таки начинать снизу:
1. Первичная конфигурация Security (без JWT)
2. репозиторий-сервис User:
3. JWT, JwtService, включение Jwt в фильтр Security, добавление Jwt в UserService
4. Контроллеры
5. Сообщения Kafka
